### DISCUSSION TITLE
[Feature Request] Cost Kill Switch

### DISCUSSION THREAD

**Discussion by p5** [runs-on/runs-on#175] [Ideas]
I'm hoping to use runs-on for an open-source project I'm involved with, though my primary worry is costs running away from me.  

I would love it if (somehow) we were able to configure a "kill switch" which prevents any new jobs from being fired off after a cost threshold has been reached. This would act as a safety net so costs stay under control.

**Comment by crohr** [comment]
This is something I’d love to have, but it is not currently possible because AWS does not provide billing granularity at the correct level / interval. I believe it should be a feature of AWS to set a max budget for an account and have it disable services if it goes above. I don’t think RunsOn can fix that by itself unfortunately. 

You do have the option to set an alarm for the consumed minutes though, so it is at least something.

**Reply by p5** [reply]
Could we fork off from the existing budgets alarms, which can (or already are) sent to an SQS/SNS queue?

When a 100%+ budget alarm is detected, perhaps we can invoke a Lambda/something else to put a "lock" value in a DynamoDB table.  That table can be checked when scheduling tasks.

I know I haven't explained it too well, so I'll try again:
1. AWS Budgets sends SNS notification to a queue
2. Runs-on monitors that queue and, when 100%+ is received, stores a "lock value" somewhere
3. When runs-on tries to schedule new tasks, if that "lock value" is detected, the workflows fail
4. On the 1st of each month, reset the lock

Now, I understand it won't be entirely accurate, since cost notifications aren't immediate, though it's better than nothing.

**Comment by crohr** [comment]
Alright maybe I need to have a closer look at what’s possible here. Thanks for the additional context!

**Comment by crohr** [comment]
I'm going to close this for now. Didn't have the need come up from other users yet, and I think a kill switch would quite confusing for developers. Thanks for the idea though!
